### PR TITLE
[Do not merge yet] Remove alloc_jemalloc, switch the default global allocator to System

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes https://github.com/rust-lang/rust/issues/36963

**Do not merge** yet. This PR by itself is likely to regress rustc performance. The purpose of opening it now is to measure by how much. We’ll likely want to figure out https://github.com/rust-lang/rust/issues/51038 and land them around the same time.

### CODE REVIEW

**Comment by SimonSapin** [comment]
@bors try

@Mark-Simulacrum, could you run this on perf.rlo?

**Comment by cuviper** [comment]
Please specifically note the versions of the system allocator (presumably glibc) when reporting performance.  If possible, it would be nice to have results for both current and older glibc, say Fedora 28 versus RHEL/CentOS 7.  If there's a way I can help generate results, let me know -- I don't have hardware to dedicate to this, but I can do manual runs for this evaluation.

**Comment by SimonSapin** [comment]
Indeed, results on perf.rust-lang.org are presumably collected on Linux with a single version of glibc. Measuring impact on macOS would be nice too.

**Comment by Mark-Simulacrum** [comment]
Perf results, once ready, will have been run with Ubuntu GLIBC 2.23-0ubuntu10.

**Comment by cuviper** [comment]
OK.  The most notable change for glibc came in 2.26:
https://sourceware.org/ml/libc-alpha/2017-01/msg00524.html

For Ubuntu, that would be in artful (17.10) and later.

**Comment by SimonSapin** [comment]
In this case, getting both sets of results would be interesting.

**Comment by SimonSapin** [comment]
So I tried running the benchmarks on a mac and got this:

```
thread 'main' panicked at 'assertion failed: has_perf', collector/src/bin/rustc-perf-collector/execute.rs:313:9
```

I assume `has_perf` refers to Linux’s perf tool. Is there a mode for the data collector to get the subset of available measures, possibly only wall clock time?

I also ran it on an Ubuntu 18.04 machine with the `libc6` package at version 2.27-3ubuntu1 and an i7 CPU with 4 cores / 8 threads at 4.6 GHz. http://localhost:2346/compare.html?start=64f8ae08fd5294718466a61f74f6de0135b88035&end=91eac4dd1e41932c605075305758d84406abfaf4 does show results, but there’s so much green (where I would have expected at least some red) that I wonder if I’m comparing data sets measured on different machines.

```
compare	2018-07-03 (64f8ae08)	2000-01-01 (91eac4dd)	% change
 
tokio-webpush-simple-opt
	avg: -22.2%	min: -65.7%	max: -0.4%
unify-linearly-opt
	avg: -16.1%	min: -26.7%	max: -12.0%
helloworld-opt
	avg: -17.2%	min: -22.2%	max: -15.1%
helloworld-debug
	avg: -16.3%	min: -17.5%	max: -15.5%
unify-linearly-debug
	avg: -13.7%	min: -16.3%	max: -12.1%
```

<details>

```
issue-46449-opt
	avg: -5.1%	min: -9.3%	max: -0.8%
coercions-opt
	avg: -5.3%?	min: -8.9%?	max: -3.2%?
unused-warnings-opt
	avg: -6.3%	min: -6.9%	max: -5.3%
unused-warnings-debug
	avg: -5.9%	min: -6.9%	max: -5.2%
coercions-debug
	avg: -4.6%?	min: -6.4%?	max: -3.6%?
html5ever-opt
	avg: -3.3%	min: -5.4%	max: -1.5%
html5ever-debug
	avg: -3.5%	min: -5.3%	max: -1.6%
html5ever-check
	avg: -3.7%	min: -5.2%	max: -1.7%
deep-vector-debug
	avg: -2.9%	min: -4.4%	max: -2.1%
deep-vector-opt
	avg: -2.3%	min: -4.4%	max: -1.4%
tuple-stress-opt
	avg: -2.3%	min: -3.8%	max: -0.0%
script-servo-debug
	avg: -1.6%	min: -3.7%	max: -0.7%
coercions-check
	avg: -2.6%?	min: -3.7%?	max: -1.7%?
tuple-stress-check
	avg: -1.9%	min: -3.3%	max: -0.0%
style-servo-debug
	avg: -2.1%?	min: -3.2%?	max: -0.9%?
tokio-webpush-simple-debug
	avg: -1.8%	min: -3.1%	max: -1.2%
tuple-stress-debug
	avg: -2.0%	min: -3.1%	max: -0.0%
sentry-cli-opt
	avg: -1.2%	min: -3.0%	max: -0.4%
inflate-opt
	avg: -1.9%	min: -2.9%	max: -1.2%
encoding-debug
	avg: -1.6%	min: -2.7%	max: -0.7%
ripgrep-opt
	avg: -1.3%	min: -2.7%	max: -0.4%
unused-warnings-check
	avg: -2.3%	min: -2.7%	max: -1.8%
style-servo-opt
	avg: -1.0%?	min: -2.7%?	max: -0.4%?
syn-debug
	avg: -1.7%	min: -2.6%	max: -1.0%
issue-46449-debug
	avg: -1.8%	min: -2.5%	max: -1.5%
ripgrep-debug
	avg: -1.5%	min: -2.5%	max: -0.6%
regression-31157-opt
	avg: -1.2%	min: -2.5%	max: -0.1%
regex-debug
	avg: -1.4%	min: -2.4%	max: -0.3%
syn-opt
	avg: -0.7%	min: -2.3%	max: 1.9%
deep-vector-check
	avg: -1.9%	min: -2.3%	max: -1.6%
regex-opt
	avg: -1.1%	min: -2.2%	max: -0.3%
sentry-cli-check
	avg: -1.9%	min: -2.1%	max: -1.5%
inflate-debug
	avg: -0.1%	min: -2.1%	max: 0.6%
encoding-check
	avg: -1.7%	min: -2.0%	max: -1.6%
regression-31157-debug
	avg: -1.3%	min: -2.0%	max: -0.7%
encoding-opt
	avg: -1.0%	min: -2.0%	max: -0.2%
ripgrep-check
	avg: -1.6%	min: -2.0%	max: -1.3%
futures-debug
	avg: -1.3%	min: -1.9%	max: -0.9%
serde-debug
	avg: -1.2%	min: -1.9%	max: -0.9%
deeply-nested-debug
	avg: -1.1%	min: -1.9%	max: -0.8%
issue-46449-check
	avg: -1.6%	min: -1.9%	max: -1.4%
piston-image-debug
	avg: -1.1%	min: -1.9%	max: -0.3%
serde-opt
	avg: -1.2%	min: -1.8%	max: -0.8%
regression-31157-check
	avg: -1.3%	min: -1.8%	max: -0.9%
sentry-cli-debug
	avg: -1.0%	min: -1.8%	max: -0.4%
regex-check
	avg: -1.7%	min: -1.8%	max: -1.5%
syn-check
	avg: -1.6%	min: -1.8%	max: -1.4%
deeply-nested-check
	avg: -1.2%	min: -1.7%	max: -0.9%
futures-opt
	avg: -1.2%	min: -1.7%	max: -0.8%
tokio-webpush-simple-check
	avg: -1.4%	min: -1.7%	max: -1.3%
cargo-opt
	avg: -0.7%	min: -1.7%	max: -0.4%
deeply-nested-opt
	avg: -0.6%	min: -1.6%	max: -0.1%
cargo-check
	avg: -1.1%	min: -1.6%	max: -0.8%
webrender-check
	avg: -1.2%	min: -1.6%	max: -1.1%
webrender-debug
	avg: -1.0%	min: -1.5%	max: -0.4%
piston-image-opt
	avg: -0.9%	min: -1.5%	max: -0.3%
style-servo-check
	avg: -1.2%?	min: -1.5%?	max: -0.9%?
clap-rs-opt
	avg: -0.8%	min: -1.4%	max: -0.5%
futures-check
	avg: -1.1%	min: -1.4%	max: -0.8%
piston-image-check
	avg: -1.3%	min: -1.4%	max: -1.1%
clap-rs-check
	avg: -0.9%	min: -1.4%	max: -0.5%
unify-linearly-check
	avg: -1.0%	min: -1.4%	max: -0.6%
cargo-debug
	avg: -0.8%	min: -1.4%	max: -0.5%
webrender-opt
	avg: -1.0%	min: -1.3%	max: -0.7%
helloworld-check
	avg: -1.0%	min: -1.3%	max: -0.6%
inflate-check
	avg: 0.7%	min: -0.9%	max: 1.3%
script-servo-check
	avg: -0.5%	min: -1.2%	max: -0.0%
serde-check
	avg: -0.8%	min: -1.0%	max: -0.5%
clap-rs-debug
	avg: -0.6%	min: -0.8%	max: -0.4%
script-servo-opt
	avg: -0.5%	min: -0.7%	max: -0.2%
```

</details>

**Comment by mati865** [comment]
@SimonSapin `http://localhost:2346` surely isn't vaild url :wink:

**Comment by SimonSapin** [comment]
The results on the infra look similar to those on my machine… and I still don’t know what to make of them. CC @rust-lang/wg-compiler-performance, @glandium, @nnethercote 

* https://perf.rust-lang.org/compare.html?start=64f8ae08fd5294718466a61f74f6de0135b88035&end=91eac4dd1e41932c605075305758d84406abfaf4
* https://perf.rust-lang.org/compare.html?start=64f8ae08fd5294718466a61f74f6de0135b88035&end=91eac4dd1e41932c605075305758d84406abfaf4&stat=wall-time

**Comment by SimonSapin** [comment]
TL;DR of the numbers: depending on what is being compiled in what mode, compilation wall clock time difference due do disabling jemalloc is between -42.5% (time nearly halved!) and +10.6%

**Comment by alexcrichton** [comment]
@SimonSapin oh the perf links you gisted actually were for a [pretty wide range of commits](https://github.com/rust-lang/rust/compare/64f8ae08fd5294718466a61f74f6de0135b88035...91eac4dd1e41932c605075305758d84406abfaf4), but [this comparison URL](https://perf.rust-lang.org/compare.html?start=860d169474acabdc53b9a698f8ce02eba7e0daeb&end=91eac4dd1e41932c605075305758d84406abfaf4&stat=wall-time) is just for the jemalloc changes. 

From that URL some conclusions are:

* Tiny programs tend to compile faster (0.17s -> 0.15s) ish
* Big programs compile a little slower (187s -> 211s for script servo debug)

Overall I think the losses here outweigh the gains (as the gains are all very small and only on small programs) and we probably want to stick with jemalloc on Linux for now. It'd be interesting though to compare with the glibc 2.26 allocator as well though!

**Comment by nnethercote** [comment]
I agree with @alexcrichton. The ones that improved were toy or stress programs; all the more interesting benchmarks regressed :(

**Comment by michaelwoerister** [comment]
I agree with @alexcrichton and @nnethercote, we should not merge this as is. Thanks a lot for measuring though, @SimonSapin!

**Comment by mati865** [comment]
For redis glibc 2.26 [can make huge difference](https://www.phoronix.com/scan.php?page=news_item&px=Glibc-2.26-Redis-Test)

It should be revaluated again once perf.rust-lang.org is updated to something more recent.

**Comment by michaelwoerister** [comment]
> For redis glibc 2.26 can make huge difference

Those numbers look great but wouldn't that mean that people on older systems would still experience a noticeable regression?

**Comment by mati865** [comment]
> Those numbers look great but wouldn't that mean that people on older systems would still experience a noticeable regression?

Unfortunately that's the case however distributions like Ubuntu, Fedora, openSUSE/SUSE already provide Rust built without jemalloc in their repositories.

There is no need to hurry but at some the focus should be shifted for more recent systems.

**Comment by steveklabnik** [comment]
> Unfortunately that's the case however distributions like Ubuntu, Fedora, openSUSE/SUSE already provide Rust built without jemalloc in their repositories.

This is an interesting point; if we’re basing this situation on “not trying to regress distros”, we may be making a category error. It’s likely they’ve already regressed, in a sense.

IMHO, we should also be taking binary size into some consideration here. It’s a persistent foot gun that people compare C with the system allocator to Rust with jemalloc.

**Comment by SimonSapin** [comment]
@steveklabnik There are people using rustc packaged by their distro, but also users of various distros with various versions of glibc using our rustc builds through rustup.

Re binary size I assume you mean changing *default* allocator for Rust programs to `System`, https://github.com/rust-lang/rust/issues/36963. I think there’s strong consensus to do that, but not to do it without also making the compiler continue to use jemalloc somehow. So I think we’ll end up merging most of the changes in this PR eventually, but likely not before fixing https://github.com/rust-lang/rust/issues/51038.

**Comment by steveklabnik** [comment]
@SimonSapin ah; yes, that was what I was talking about; I got my threads mixed up. Thanks.

**Comment by SimonSapin** [comment]
> So I think we’ll end up merging most of the changes in this PR eventually, but likely not before fixing #51038.

There’s no point keeping this PR open in the meantime, closing to make triage easier. Please direct conversation to #51038 or #36963 as appropriate.
